File: leetcode-cn/0429.0_N-ary_Tree_Level_Order_Traversal.py
'''
approach: BFS
Time: O(N)
Space: O(N)

执行用时：44 ms, 在所有 Python 提交中击败了48.84%的用户
内存消耗：15.9 MB, 在所有 Python 提交中击败了64.53%的用户
'''


"""
# Definition for a Node.
class Node(object):
    def __init__(self, val=None, children=None):
        self.val = val
        self.children = children
"""

class Solution(object):
    def levelOrder(self, root):
        """
        :type root: Node
        :rtype: List[List[int]]
        """
        if not root:
            return []
        
        q = [root]
        level = []
        while q:
            current_level_values = []
            next_level = []
            for node in q:
                for child in node.children:
                    next_level.append(child)
                current_level_values.append(node.val)
            level.append(current_level_values)
            q = next_level

        return level


'''
approach: BFS
T: O(N)
S: O(N)

执行用时：52 ms, 在所有 Python3 提交中击败了82.93% 的用户
内存消耗：16.8 MB, 在所有 Python3 提交中击败了76.37% 的用户
通过测试用例：38 / 38
'''
"""
# Definition for a Node.
class Node:
    def __init__(self, val=None, children=None):
        self.val = val
        self.children = children
"""

class Solution:
    def levelOrder(self, root: 'Node') -> List[List[int]]:
        if not root:
            return []
        ans = []
        q = [root]
        while q:
            level = []
            next_q = []
            for node in q:
                level.append(node.val)
                next_q.extend(node.children)
            ans.append(level)
            q = next_q
            
        return ans 


'''
DFS

执行用时：56 ms, 在所有 Python3 提交中击败了65.17% 的用户
内存消耗：17 MB, 在所有 Python3 提交中击败了29.48% 的用户
通过测试用例：38 / 38
'''
"""
# Definition for a Node.
class Node:
    def __init__(self, val=None, children=None):
        self.val = val
        self.children = children
"""

class Solution:
    def levelOrder(self, root: 'Node') -> List[List[int]]:
        ans = []

        def dfs(node: Node, depth: int) -> None:
            if not node:
                return 
            if len(ans) <= depth:
                ans.append([])
            ans[depth].append(node.val)
            for ch in node.children:
                dfs(ch, depth + 1)
            
        dfs(root, 0)
        return ans 

File: cpp/0429.0_N-ary_Tree_Level_Order_Traversal.cpp
/*
BFS

执行用时：16 ms, 在所有 C++ 提交中击败了79.75% 的用户
内存消耗：11.7 MB, 在所有 C++ 提交中击败了18.26% 的用户
通过测试用例：38 / 38
*/
/*
// Definition for a Node.
class Node {
public:
    int val;
    vector<Node*> children;

    Node() {}

    Node(int _val) {
        val = _val;
    }

    Node(int _val, vector<Node*> _children) {
        val = _val;
        children = _children;
    }
};
*/

class Solution {
public:
    vector<vector<int>> levelOrder(Node* root) {
        vector<vector<int>> ans;
        if (!root) return ans;
        queue<Node*> q;
        q.push(root);
        while (!q.empty()) {
            int n = q.size();
            vector<int> level;
            while (n--) {
                Node *p = q.front();
                q.pop();
                level.push_back(p->val);
                for (int i=0; i < p->children.size(); i++) {
                    q.push(p->children[i]);
                }
            }
            ans.push_back(level);
        }
        return ans;
    }
};





/*
BFS

执行用时：12 ms, 在所有 C++ 提交中击败了97.03% 的用户
内存消耗：11.5 MB, 在所有 C++ 提交中击败了84.51% 的用户
通过测试用例：38 / 38
*/
/*
// Definition for a Node.
class Node {
public:
    int val;
    vector<Node*> children;

    Node() {}

    Node(int _val) {
        val = _val;
    }

    Node(int _val, vector<Node*> _children) {
        val = _val;
        children = _children;
    }
};
*/

class Solution {
public:
    vector<vector<int>> levelOrder(Node* root) {
        if (nullptr == root) return {};
        queue<Node*> q;
        vector<vector<int>> ans;
        q.push(root);
        while (!q.empty()) {
            int n = q.size();
            vector<int> level;
            for (int i = 0; i < n; i++) {
                Node *p = q.front();
                q.pop();
                level.push_back(p->val);
                for (auto ch: p->children) {
                    q.push(ch);
                }
            }
            ans.push_back(level);
        }
        return ans;
    }
};


/*
执行用时：12 ms, 在所有 C++ 提交中击败了97.03% 的用户
内存消耗：11.5 MB, 在所有 C++ 提交中击败了80.82% 的用户
通过测试用例：38 / 38
*/
/*
// Definition for a Node.
class Node {
public:
    int val;
    vector<Node*> children;

    Node() {}

    Node(int _val) {
        val = _val;
    }

    Node(int _val, vector<Node*> _children) {
        val = _val;
        children = _children;
    }
};
*/

class Solution {
public:
    vector<vector<int>> levelOrder(Node* root) {
        if (nullptr == root) return {};
        queue<Node*> q;
        vector<vector<int>> ans;
        q.push(root);
        while (!q.empty()) {
            int n = q.size();
            ans.push_back(vector<int>());
            for (int i = 0; i < n; i++) {
                Node *p = q.front();
                q.pop();
                ans.back().push_back(p->val);
                for (auto& ch: p->children) {
                    q.push(ch);
                }
            }
        }
        return ans;
    }
};



